### PR TITLE
visit: add v3.4.0, v3.4.1

### PR DESCRIPTION
**Pull request by kwryankrattiger**
Add the 3.4 versions to Spack to get upgraded VTK deps and therefore upgraded python deps.

### CODE REVIEW

**Comment by kwryankrattiger** [comment]
@cyrush I was able to get this building after some effort.

Do you know how/why VisIt is looking for the library `hdf5_mpi`? Looking at the MPI enabled installs of HDF5 1.12 and 1.14, this library does not exist. I think possible VisIt is building HDF5 1.8 or older by default? Post ECP, exa-scale era, I think we want to move to 1.14 as much as possible for the VOL adapters implemented there. The only way to saw not to search for that library was to disable `VISIT_PARALLEL`, which is not desired. The other way, which I did here, was to have HDF5 mock installing the `_mpi` version with a symlink.

**Comment by cyrush** [comment]
Most use of HDF5 in VisIt does not require MPI. However there are some plugins that depend on it.  

Due to this, we have support for linking and using both a non-mpi and an mpi build of HDF5 in a single VisIt build.  Some combo of plugins is expecting this and I think the only way to get there is via our build_visit script. We may have to disable the plugin that requires this. 

We are going to move to support an mpi-enabled hdf5 even for serial builds, but that won't be until a future release. 

For updating to 1.14, we are working on vetting performance and compatibility implications of updating. 
Mark Miller is working on a detailed look at this. 

Thank you for helping with this.

I find it very hard to navigate the spack ci outputs, some much info and hard to parse what options are in play.

**Comment by kwryankrattiger** [comment]
> For updating to 1.14, we are working on vetting performance and compatibility implications of updating.
Mark Miller is working on a detailed look at this.

FWIW, I just built using HDF5 1.14 locally (and hopefully in CI soon!). This is possible in Spack as we are injecting API compatibility flags into the build flags for SILO and VisIt, and that will likely be the default build by the SDK going forward as v1.14 is the target HDF5 version for the SDK. Do you foresee any issues with that? I currently don't have much of a test suite to verify VisIt builds with.

I would be interested to see what kinds of performance results you obtain during testing between HDF5 1.12 and 1.14.

**Comment by cyrush** [comment]
Thanks, I don't foresee any issues with using 1.14.

We have tested and used it in several places - LLNL we are being extra cautious b/c most sim codes leverage hdf5 and we don't want surprises.

**Comment by bernhardkaindl** [comment]
On [Adios2: Restrict python, 3.11 did not work](https://github.com/spack/spack/pull/47161/commits/11905464c5ce85de3f5608b01d0fe1dcca24fc12): I merged #47219 that prohibits `Python@3.11` with `adios2@:2.7`

I found that not even Python@3.8 works with `adios2@:2.7`,
and proposed to limit `+python` to `@2.8:` in the variant itself as part of my suggestions in:
- #47235

This is just to say: [Adios2: Restrict python, 3.11 did not work](https://github.com/spack/spack/pull/47161/commits/11905464c5ce85de3f5608b01d0fe1dcca24fc12) is now obsoleted by the merged #47219, but if you keep it in the PR, it does not prevent the merge.

**Comment by kwryankrattiger** [comment]
@bernhardkaindl I would prefer preserving python bindings as much as possible. As great a language as c++ is, people seem to prefer using the python bindings and the DAV really wants to make that available whenever possible, even if it is a little difficult.

@cyrush I was looking at the release notes and I didn't see an adios2 version bumb at all. What is the latest usable with 3.4?

**Comment by cyrush** [comment]
VisIt 3.4.1 was released using ADIOS-2 2.10.0-rc1

**Comment by cyrush** [comment]
VisIt 3.4.0 was released using ADIOS-2 2.7.1

**Comment by kwryankrattiger** [comment]
Great. Judging from the comments the only restriction was related to https://github.com/visit-dav/visit/issues/19209 which is fixed. I noted that in the the recipe.

**Comment by kwryankrattiger** [comment]
@cyrush  The last error I am seeing is `visit +gui` seems to need `py-pip` to build the python module. Is this really a GUI only requirement or is the python build in the non-GUI specs not building the python module correctly?

**Comment by cyrush** [comment]
@kwryankrattiger we should not need pip for the gui, I expect that is somehow related to the cli (python interface)

**Comment by kwryankrattiger** [comment]
Okay, I think I see the issue now. VisIt is not buiding the python CLI unless the GUI is enabled. See v3.4.1 `src/CMakeLists.txt:1210`

```
if(NOT VISIT_SERVER_COMPONENTS_ONLY AND NOT VISIT_ENGINE_ONLY AND NOT VISIT_DBIO_ONLY)
    add_subdirectory(vtkqt)
    add_subdirectory(winutil)
    add_subdirectory(gui)
    if(VISIT_PYTHON_SCRIPTING AND PYTHONLIBS_FOUND)
        if(NOT VISIT_STATIC)
            add_subdirectory(visitpy)
        endif()
    endif()
    # always add java subdir, so code gen targets can be active even
    # if java is not enabled
    add_subdirectory(java)
endif()
```

It seems the only way to get the CLI without the GUI is to build `VISIT_STATIC`, which we explicitly do not want to do in the SDK as Shared libraries are preferred as the default.

@cyrush Is requiring the GUI for the python CLI a bug or intended behavior?

The other issue is not getting the `py-pip` module as a dependency when `+python`. This needs to be properly checked in the VisIt configuration step as well, currently there is no check for pip. It is used later in `PYTHON_ADD_PIP_SETUP` where we add a custom command to build the python extensions.

This issue has gone unnoticed for a while it seems as we have not built the VisIt GUI and/or tested the Python extensions in Spack installs.

**Comment by BenWibking** [comment]
Since Visit supports qt6 now, is there any chance the Spack package will support qt6 for `+gui` builds?

This is required for VisIt spack builds to work on macOS now, since qt5 no longer builds on macOS.

**Comment by kwryankrattiger** [comment]
@BenWibking  Qt6 support will come later. My goal in this is to quickly get VisIt in a state where it is buildable in Spack CI after #46913.

Feel free to open an issue and link it here. We are also working on moving other VTK 9 apps, ref .ParaView  #45300, to Qt6.

**Comment by adamjstewart** [comment]
Any updates on this? Would like to deprecate Python 3.8 before the next Spack release

**Comment by kwryankrattiger** [comment]
Hoping to finish up today or tomorrow.

**Comment by kwryankrattiger** [comment]
We get everything to the install step, and then it fails. I can reproduce this locally and it seems to be somehow related to using a `MODULE` instead of a `SHARED_LIBRARY` type for the `visitmodule` target, changing the type in the install script fixes the install. Patching that doesn't seem like a satisfying answer to this though, there is likely something else going on that I am missing still.

**Comment by kwryankrattiger** [comment]
@scottwittenburg found a similar installation issue in another project which was temporarily resolved by using the makefile generator. Testing locally this seems to have also resolved the VisIt installation issue. As a quick fix putting in a conflict for using Ninja with the python modules enabled.

**Comment by kwryankrattiger** [comment]
The issue with the VisIt install is being investigated and tracked here https://gitlab.kitware.com/cmake/cmake/-/issues/26445

@cyrush It is not yet clear to me if this is a CMake or Pip or VisIt/Conuit issue. Currently the workaround is to use the Makefile generator when building python extensions for VisIt but hopefully we will be able to revert that soon. CI seems like it will pass now so we should be able to unblock #46913

File: var/spack/repos/builtin/packages/hdf5/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/hdf5/package.py]: What about macOS? Do we need to use `dso_suffix`?

**Comment by adamjstewart** [comment]
@spackbot maintainers

File: var/spack/repos/builtin/packages/hdf5/package.py
Comment (kwryankrattiger) [reply var/spack/repos/builtin/packages/hdf5/package.py]: Good call, updated to use `find_libraries` instead to capture this correctly on all systems.

**Review by hyoklee — APPROVED** [review]

**Comment by BenWibking** [comment]
I tried building this on my Fedora 41 laptop and I am getting a build failure for Visit related to finding the HDF5 MPI variants:
```
     255    -- Looking for HDF5
     256    --   Found library hdf5 in /home/benwibking/spack/opt/spack/linux-fedora_asahi_remix41-m1/gcc-12.4.0/hdf5-1.14.5-c
            57vtzrhqnuhdog3bawtcj3r2k7s3uor/lib
     257    --   HDF5 found
     258    -- Looking for HDF5_MPI
  >> 259    CMake Error at CMake/SetUpThirdParty.cmake:223 (message):
     260      Library hdf5_mpi not found in
     261      /home/benwibking/spack/opt/spack/linux-fedora_asahi_remix41-m1/gcc-12.4.0/hdf5-1.14.5-c57vtzrhqnuhdog3bawtcj3r2k
            7s3uor/lib
     262    Call Stack (most recent call first):
     263      CMake/FindHDF5.cmake:44 (SET_UP_THIRD_PARTY)
     264      CMake/SetUpThirdParty.cmake:588 (include)
     265      CMakeLists.txt:928 (include)
```

I used the spec `visit@3.4.1~mfem~netcdf~conduit%gcc@12.4.0`.

```
* **Spack:** 0.23.0.dev0 (c6ad47eb2c3893a3afd8c3f9e9debea1d3fd2c43)
* **Python:** 3.13.0
* **Platform:** linux-fedora_asahi_remix41-m1
```

The full build log is here: 
[visit-spack-build-out.txt](https://github.com/user-attachments/files/17793075/visit-spack-build-out.txt)

**Comment by BenWibking** [comment]
Maybe a new variant should be added for the `hdf5_mpi` shared libraries, so that it doesn't pick up pre-existing HDF5 installs?

**Comment by BenWibking** [comment]
Running `spack uninstall --dependents hdf5` and re-installing Visit fixed that problem.

Now it crashes on startup:
```console
$ visit
Running: gui3.4.1
Qt: Warning: Could not find the Qt platform plugin "wayland" in ""
Running: viewer3.4.1 -geometry 769x945+743+0 -borders 26,4,4,4 -shift 0,0 -preshift 4,26 -defer -host 127.0.0.1 -port 5600
Running: mdserver3.4.1 -host 127.0.0.1 -port 5601
VisIt's viewer exited abnormally! Aborting the Graphical User Interface. VisIt's developers may be reached via our GitHub discussions, https://github.com/visit-dav/visit/discussions
```

**Comment by BenWibking** [comment]
Is there a missing dependency here? `spack install visit@3.3.3+gui` works on this machine.
Running `visit -debug 5` shows:
```
Qt: Warning: qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 3.2, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize 0, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize 0, stencilBufferSize 0, samples 0, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::CoreProfile)
Qt: Warning: No XVisualInfo for format QSurfaceFormat(version 3.2, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize 0, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize 0, stencilBufferSize 0, samples 0, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::CoreProfile)
Qt: Warning: Falling back to using screens root_visual.
Qt: Warning: qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 3.2, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize 0, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize 0, stencilBufferSize 0, samples 0, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::CoreProfile)
Qt: Warning: qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 3.2, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize 0, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize 0, stencilBufferSize 0, samples 0, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::CoreProfile)
Qt: Fatal: Could not initialize GLX
signalhandler_core: SIGABRT! (Aborted)
```

**Comment by kwryankrattiger** [comment]
@BenWibking 

> Is there a missing dependency here?

Maybe? Did you build with GLX? What is the GL provider in your concrete environment?

**Comment by kwryankrattiger** [comment]
@BenWibking 

> Maybe a new variant should be added for the hdf5_mpi shared libraries, so that it doesn't pick up pre-existing HDF5 installs?

Adding variants for temporary fixes due to bugs in package configs is _not_ something that I would encourage...we would have a ton of useless variants all of the place if we did that. Packages should have variants that express package functionality, not build hacks.

We could also have maybe fixed this with a patch to VisIt, but to do that would have been significantly more effort. VisIt implements its own logic for finding HDF5, and other packages, and it is not version aware to the HDF5 it finds. It simply wants HDF5 and then relies on the user to guarantee the correct API is used (ref. https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/visit/package.py#L222)

**Comment by kwryankrattiger** [comment]
@BenWibking 

> Now it crashes on startup:

It appears that Spack Qt ignores wayland. I am not sure what kind of support guarantees are there for wayland, but I have never built any Spack UI packages and run them in a Wayland environment. Was this working before? How? Are you sure you have the same stack as before, if so share more about how you are building?

https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/qt/package.py#L777

**Comment by BenWibking** [comment]
> It appears that Spack Qt ignores wayland. I am not sure what kind of support guarantees are there for wayland, but I have never built any Spack UI packages and run them in a Wayland environment. Was this working before? How? Are you sure you have the same stack as before, if so share more about how you are building?
> 
> https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/qt/package.py#L777

It was working before in GNOME Wayland on Fedora 39 and Fedora 40 with Visit 3.3.3 and an older Spack develop commit (I don't know which one, sorry).

I just tried to rebuild Visit 3.3.3 with this Spack PR on Fedora 41 (w/ Wayland) in order to try to reproduce, but unfortunately I now get a build failure:
```

2 errors found in build log:
     107    -- Looking for include file sys/stat.h - found
     108    -- Looking for include file unistd.h
     109    -- Looking for include file unistd.h - found
     110    -- Looking for include file stdbool.h
     111    -- Looking for include file stdbool.h - found
     112    -- Checking for VTK in /home/benwibking/spack/opt/spack/linux-fedora_asahi_remix41-m1/gcc-12.4.0/vtk-8.2.1a-fdntdd
            p3i4lty5rf2jtodyrhznkw4jum
  >> 113    CMake Error at CMake/FindVisItVTK.cmake:109 (find_package):
     114      find_package called with invalid argument "8a.2a"
     115    Call Stack (most recent call first):
     116      CMakeLists.txt:1268 (INCLUDE)
     117
     118
     119    ERRORVisIt requires vtkCommonCore

     ...

     138    --   VTK_INCLUDE_DIRS=
     139    --   VTK_DEFINITIONS=
     140    --   VTK_LIBRARIES=
     141    --   VTK_INSTALL_PREFIX=
     142    --   VTK_LIBRARY_DIRS=/lib
     143    --   VTK_PY_WRAPPERS_DIR=/lib/python3.12;/lib/python3.13
  >> 144    CMake Error at CMake/FindVisItVTK.cmake:214 (IF):
     145      if given arguments:
     146
     147        "EXISTS" "/lib/python3.12" "/lib/python3.13/site-packages/vtk"
     148
     149      Unknown arguments specified
     150    Call Stack (most recent call first):
```

I have no externals defined. I did have to bootstrap gcc 12 on this machine, since Fedora 41 ships with gcc 14, which can't build Qt5 or VTK.

Is the OpenGL provider a new feature? I just noticed this in the docs: https://spack.readthedocs.io/en/latest/gpu_configuration.html#using-an-external-opengl-api. I wonder if previously it worked because it was actually using osmesa...

**Comment by BenWibking** [comment]
I don't think the Qt support for Wayland is an issue. It should just transparently use XWayland. And I can see that the GUI launches (it shows up onscreen). The OpenGL viewer just crashes (which then takes down the other processes).

**Comment by BenWibking** [comment]
I think the issue is that 3.4.1 should be using osmesa (since I don't have an opengl external defined), but it isn't.

File: var/spack/repos/builtin/packages/visit/package.py
Comment (kwryankrattiger) [inline var/spack/repos/builtin/packages/visit/package.py]: This breaks the 3.3.3 version of VisIt when using the unofficial patched version VTK 8.2.1a

**Comment by kwryankrattiger** [comment]
> I think the issue is that 3.4.1 should be using osmesa (since I don't have an opengl external defined), but it isn't.

1. I think you need GLX to use the GUI, I haven't recently tried but I don't think OSMesa works with Qt render windows. I could be wrong, let me know.
```python
      conflicts(
      ¦   "+gui", when="^[virtuals=gl] osmesa", msg="GUI cannot be activated with OSMesa front-end"
      )
```

3. Mesa can build GLX, you just need to specify `^[virtuals=gl] glx ^mesa+glx`

**Comment by BenWibking** [comment]
>     3. Mesa can build GLX, you just need to specify `^[virtuals=gl] glx ^mesa+glx`

I rebuilt with:
```
spack install visit@3.4.1~mfem~netcdf~conduit%gcc@12.4.0 "^[virtuals=gl]" glx ^mesa+glx
```
but I get the same GLX error at runtime :(

**Comment by BenWibking** [comment]
If I try to set an external OpenGL provider and build mesa with GLX (not sure if this actually makes sense?), I get an concretizer internal error:
```console
$ spack install visit@3.4.1~mfem~netcdf~conduit%gcc@12.4.0 "^[virtuals=gl]" glx ^mesa+glx
==> Error: Spack concretizer internal error. Please submit a bug report and include the command, environment if applicable and the following error message.
    visit@3.4.1%gcc@12.4.0~conduit~mfem~netcdf ^[virtuals=gl] glx ^mesa+glx is unsatisfiable, errors are:
    internal_error("Imposing constraint outside of imposed_nodes")
    internal_error("nodes in root condition set must be associated with root")
```

With the original spec (no +glx options), it fails due to `libEGL.so` being located at `/usr/lib64` instead of `/usr/lib`:
```console
$ ls /usr/include/EGL
eglext.h  egl.h  eglplatform.h
$ ls /usr/lib64/libEGL.so
/usr/lib64/libEGL.so
```

```
==> Error: Failed to install egl due to InstallError: This package is intended to be a placeholder for
        system-provided EGL(OpenGL) libraries from hardware vendors.  Please
        download and install EGL drivers/libraries for your graphics
        hardware separately, and then set that up as an external package.
        An example of a working packages.yaml:

        packages:
          egl:
            buildable: False
            externals:
            - spec: egl@1.5.0
              prefix: /usr/

        In that case, /usr/ should contain these two folders:

        include/EGL/      (egl headers, including "egl.h")
        lib               (egl libraries, including "libEGL.so")
```

**Comment by BenWibking** [comment]
Thanks for the fix. Visit 3.3.3 builds now with this PR.

The GUI + OpenGL rendering window do work with the VisIt 3.3.3 that I just built (no externals defined):
```console
$ spack install visit@3.3.3%gcc@12.4.0
[deleted]
$ spack load visit@3.3.3
$ visit
Running: gui3.3.3
Qt: Warning: Could not find the Qt platform plugin "wayland" in ""
Running: viewer3.3.3 -geometry 769x945+743+0 -borders 26,4,4,4 -shift 0,0 -preshift 4,26 -defer -host 127.0.0.1 -port 5600
Running: mdserver3.3.3 -host 127.0.0.1 -port 5601
```

Is there any runtime diagnostic I can try to figure out why 3.3.3 works but 3.4.1 doesn't?

**Comment by kwryankrattiger** [comment]
> Is there any runtime diagnostic I can try to figure out why 3.3.3 works but 3.4.1 doesn't?

@cyrush may have an idea of what is going on here.

**Comment by cyrush** [comment]
I don't understand the Qt mystery. 
This is with the same Qt version used for the 3.3.3 and the 3.4.1 build?

The big difference is VTK 8 vs VTK 9. Is the GUI itself failing to show?

For Visit 3.4.1, we are using Q6 for all for all platforms except for IBM P9 systems, which have very old system x11 support.

**Comment by kwryankrattiger** [comment]
ParaView is on a similar VTK  and I am not aware of any ParaView specific hacks to get Qt5 working. My suspicion is that whatever GLX is getting pulled in is not happy.

Make sure you are pulling in the GL you expect. Is is possible that the newer VTK is pointing at the wrong GL somehow (I am still trying to figure this one out for some other environments).

Running `LD_DEBUG=libs visit` should dump all of that information

**Comment by BenWibking** [comment]
> I don't understand the Qt mystery. This is with the same Qt version used for the 3.3.3 and the 3.4.1 build?
> 
Yes, same Qt5 version for both.

> The big difference is VTK 8 vs VTK 9. Is the GUI itself failing to show?
> 
No, the GUI shows in both cases (although briefly for 3.4.1 before quitting). The viewer process fails to initialize GLX, and then it all quits.

**Comment by cyrush** [comment]
> > I don't understand the Qt mystery. This is with the same Qt version used for the 3.3.3 and the 3.4.1 build?
> 
> Yes, same Qt5 version for both.
> 
> > The big difference is VTK 8 vs VTK 9. Is the GUI itself failing to show?
> 
> No, the GUI shows in both cases (although briefly for 3.4.1 before quitting). The viewer process fails to initialize GLX, and then it all quits.

Ok, something is crossed with GL init that is different for the VTK 9 case. 

If you have strace installed -- for a firehose of info that might pinpoint if two GLs are in play, you can try:

```
bin/visit -strace viewer
```

**Comment by BenWibking** [comment]
> Ok, something is crossed with GL init that is different for the VTK 9 case.
> 
> If you have strace installed -- for a firehose of info that might pinpoint if two GLs are in play, you can try:
> 
> ```
> bin/visit -strace viewer
> ```

This turned up something interesting.

Both 3.3.3 and 3.4.1 end up loading `/lib64/libGLX.so.0` (outside of the Spack tree, even though I have no externals defined), but 3.3.3 loads it after loading a lot of other shared libraries, whereas 3.4.1 loads it relatively early. I will keep looking though the strace output to see if there's anything else here.

Here is the full strace output for both versions: [visit_strace_output.tar.gz](https://github.com/user-attachments/files/17807773/visit_strace_output.tar.gz)

**Comment by BenWibking** [comment]
For 3.4.1, it also loads `libGLX_mesa.so` and `libGLX_system.so` multiple times (this is not done by 3.3.3):
```console
$ grep "openat" visit_333_strace_output.txt | grep -v ENOENT | grep libGLX
1731976854.045266 openat(AT_FDCWD, "/lib64/libGLX.so.0", O_RDONLY|O_CLOEXEC) = 0 <0.000030>
$ grep "openat" visit_341_strace_output.txt | grep -v ENOENT | grep libGLX
1731976780.102615 openat(AT_FDCWD, "/lib64/libGLX.so.0", O_RDONLY|O_CLOEXEC) = 0 <0.000014>
1731976780.845462 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000019>
1731976780.880291 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000008>
1731976780.910777 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000006>
1731976780.946628 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000013>
1731976780.997128 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000005>
1731976781.049242 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000006>
1731976781.096493 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000007>
1731976781.150859 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000008>
1731976781.200139 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000010>
1731976781.250446 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000006>
1731976781.308497 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000027>
1731976781.362212 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000007>
1731976781.411648 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000008>
1731976781.464091 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000011>
1731976781.517297 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000007>
1731976781.571510 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000009>
1731976781.621241 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000006>
1731976781.673205 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000012>
1731976781.721874 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000006>
1731976781.768618 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000010>
1731976781.805850 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000009>
1731976781.855230 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000007>
1731976781.901283 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000006>
1731976781.947678 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000007>
1731976781.989034 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000007>
1731976782.036619 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000007>
1731976782.076147 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000007>
1731976782.115175 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000010>
1731976782.157826 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000006>
1731976782.207025 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000046>
1731976782.250361 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000037>
1731976782.299542 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000009>
1731976782.337574 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000006>
1731976782.395467 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000007>
1731976782.442807 openat(AT_FDCWD, "/lib64/libGLX_mesa.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000038>
1731976782.487079 openat(AT_FDCWD, "/lib64/libGLX_system.so.0", O_RDONLY|O_CLOEXEC) = 7 <0.000007>
```

**Comment by BenWibking** [comment]
Does Visit `dlopen` all of its shared libraries? Or can https://github.com/spack/spack/pull/47365 catch shared libraries linked outside of the Spack tree in this case?

**Comment by cyrush** [comment]
@BenWibking  VisIt does not dlopen all is libs, but it does control the library paths during its launch process.

I will compare your trace to another linux system and see what I learn.

Note: VTK has completely reworked how GL loading works. When, for VisIt 3.5 we expect to use a newer version of VTK with these changes. There is hope the new system will greatly simply GL for both onscreen and offscreen together.

**Comment by adamjstewart** [comment]
Should we merge because CI is passing? Could also set the preferred version to the older version if we're uncertain about its stability.

**Comment by kwryankrattiger** [comment]
I think this is safe to merge as is.

The GL issue I think is another bug that I am investigating. I still need to make the issue, I think the investigation here was really helpful for that effort. It seems like we are in fact getting multiple GLs in the runtime, one of which is ignoring the rpath version. Even if you change the preferred version, we will still build the latest unless you say to allow deprecated due to the python requirement. Adding preferred doesn't note that the current build may not work. I think if you build with the `opengl` GL provider it will probably not have any issue.

I have a follow up patch related to the make vs ninja hack in this PR. It is stable enough right now so I don't think we need to wait for that.

**Review by adamjstewart — APPROVED** [review]
Everything looks fine to me, but maybe it would be good to get approval from @cyrush before merging.

**Review by cyrush — APPROVED** [review]

**Comment by BenWibking** [comment]
> The GL issue I think is another bug that I am investigating. I still need to make the issue, I think the investigation here was really helpful for that effort. It seems like we are in fact getting multiple GLs in the runtime, one of which is ignoring the rpath version. Even if you change the preferred version, we will still build the latest unless you say to allow deprecated due to the python requirement. Adding preferred doesn't note that the current build may not work. I think if you build with the `opengl` GL provider it will probably not have any issue.

@kwryankrattiger Is there a Github issue that I can follow for progress on the GL issue?

**Pull request merged**
